sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala

```diff
@@ -772,6 +780,13 @@ private[sql] object ParquetSchemaConverter {
   val EMPTY_MESSAGE: MessageType =
     Types.buildMessage().named(ParquetSchemaConverter.SPARK_PARQUET_SCHEMA_NAME)
 
+  // Used to annotate as metadata on the struct that replaces a VariantType when shredding.
+  val VARIANT_WRITE_SHREDDING_KEY: String = "__VARIANT_WRITE_SHREDDING_KEY"
+
+  def isVariantShreddingStruct(s: StructType): Boolean = {
+    s.fields.length > 0 && s.fields.forall(_.metadata.contains(VARIANT_WRITE_SHREDDING_KEY))
```

Comment:
Every field contains the shredding key? What happens only some have it?

Reply:
I put it on all or none in `updateSchemaForVariantShredding`, so that shouldn't really happen. We could check and fail here if there's an inconsistency. I could also just put it on the first field in the struct, if you prefer. I don't think there's any real need to put it on all of them.